azurelinuxagent/common/cgroups.py

```diff
@@ -0,0 +1,698 @@
+
+        :param path: str
+        :param hierarchy: str
+        """
+        pass
+
+    @staticmethod
+    def _setup_wrapper_groups():
+        """
+        For each hierarchy, construct the wrapper cgroup and apply the appropriate limits
+        """
+        for hierarchy in HIERARCHIES:
+            root_dir = CGroups.construct_custom_path_for_hierarchy(hierarchy)
+            CGroups._try_mkdir(root_dir)
+            CGroups._apply_wrapper_limits(root_dir, hierarchy)
+
+    @staticmethod
+    def setup():
+        """
+        Only needs to be called once, and should be called from the -daemon instance of the agent.
+            Mount the cgroup fs if necessary
+            Create wrapper cgroups for agent-plus-extensions and set limits on them;
+            Add this process to the "agent" cgroup, if required
+        Actual collection of metrics from cgroups happens in the -run-exthandlers instance
+        """
+        cgroups_enabled = False
+        try:
+            from azurelinuxagent.common import osutil
+            osutil.get_osutil().mount_cgroups()
+            CGroups._setup_wrapper_groups()
+            pid = int(os.getpid())
+            if not CGroups.is_systemd_manager():
+                cg = CGroups.for_extension(AGENT_NAME)
+                cg.add(pid)
+                logger.info("Add daemon process pid {0} to {1} cgroup".format(pid, cg.name))
+            else:
+                logger.info("Daemon process pid {0} cgroup managed by systemd".format(pid))
+            cgroups_enabled = True
+            status = "OK"
+        except CGroupsException as cge:
+            status = cge.msg
+
+        from azurelinuxagent.common.event import add_event, WALAEventOperation
+        add_event(
+            AGENT_NAME,
+            version=CURRENT_VERSION,
+            op=WALAEventOperation.InitializeCGroups,
+            is_success=cgroups_enabled,
+            message=status,
+            log_event=False)
+
+    @staticmethod
+    def add_to_extension_cgroup(name, pid=int(os.getpid())):
+        """
+        Create cgroup directories for this extension in each of the hierarchies and add this process to the new cgroup.
+        Should only be called when creating sub-processes and invoked inside the fork/exec window. As a result,
+        there's no point in returning the CGroups object itself; the goal is to move the child process into the
+        cgroup before the new code even starts running.
+
+        :param str name: Short name of extension, suitable for naming directories in the filesystem
+        :param int pid: Process id of extension to be added to the cgroup
+        """
+        if not CGroups.enabled():
+            return
+        if name == AGENT_NAME:
+            logger.warn('Extension cgroup name cannot match agent cgroup name({0})'.format(AGENT_NAME))
+
+        try:
+            logger.info("Move process {0} into cgroups for extension {1}".format(pid, name))
```

Comment:
Some (most?) of these info statements feel verbose to me.  Have you reviewed a BVT run to see how verbose they are (or are not)?

Reply:
There's no other historical record of what processes were associated with any particular cgroup. Any extension developer would need to see this information if investigating "Why did some of my resource consumption not get tracked".

Reply:
When adding an extension, this generates two events for the new extension (once for invoking with -install and once for -enable) plus one for each existing extension (the -enable that every active extension gets upon moving to a new goal state). This is roughly a 10%-20% increase in the amount of log spam every time there's a new goal state. In the steady state (when no goal state changes occur), no net new log entries are created.